File: buluo/backend/app/parser/coc_parser.py
"""
CoC JSON 解析器
================

解析逻辑:
  - JSON 中每个对象如果有 "timer" 字段, 且值 > 0, 则说明正在升级
  - 对 helpers 类别, 使用 "helper_cooldown" 字段而非 "timer"
  - finish_time = 当前时间 + timer (秒)
"""

import json
from datetime import datetime, timedelta, timezone
from pathlib import Path
from typing import Any

# ID 映射缓存: { "buildings": {"1000007": {...}}, "spells": {...}, ... }
_map_cache: dict[str, dict[str, dict]] | None = None
_MAP_PATH = Path(__file__).resolve().parent.parent / "assets" / "building_map.json"


def load_id_map() -> dict[str, dict[str, dict]]:
    """加载 ID 映射表 (带缓存), 返回按 category 分层的嵌套 dict"""
    global _map_cache
    if _map_cache is not None:
        return _map_cache

    with open(_MAP_PATH, "r", encoding="utf-8") as f:
        _map_cache = json.load(f)
    return _map_cache


def resolve_name(category: str, data_id: Any) -> str:
    """根据 category + data_id 查名称"""
    id_map = load_id_map()
    key = str(data_id)

    # 1. 精确 category 查找
    cat_map = id_map.get(category)
    if cat_map and key in cat_map:
        return cat_map[key].get("name", f"{data_id}")

    # 2. 跨 category 搜索 (equipment 特殊: 可能在不同 category 中同名)
    for cat, items in id_map.items():
        if key in items and cat != category:
            return items[key].get("name", f"{data_id}")

    return f"{data_id}"


class UpgradeItem:
    """解析后的单个升级项"""

    def __init__(
        self,
        category: str,
        item_name: str,
        item_level: int,
        timer_seconds: int,
        finish_time: datetime,
        data_id: int | None = None,
    ):
        self.category = category
        self.item_name = item_name
        self.item_level = item_level
        self.timer_seconds = timer_seconds
        self.finish_time = finish_time
        self.data_id = data_id


def _get_timer(item: dict) -> int | None:
    """从字典中获取有效的计时器值"""
    timer = item.get("timer")
    if timer is not None and isinstance(timer, (int, float)) and timer > 0:
        return int(timer)

    cooldown = item.get("helper_cooldown")
    if cooldown is not None and isinstance(cooldown, (int, float)) and cooldown > 0:
        return int(cooldown)

    return None


def _get_level(item: dict) -> int:
    """从字典中获取等级, 兼容 lvl / cnt / level 等字段名"""
    for field in ("lvl", "level"):
        val = item.get(field)
        if val is not None:
            try:
                return int(val)
            except (TypeError, ValueError):
                continue
    return 1


def _parse_category_with_timer(raw_list: list, category: str) -> list[UpgradeItem]:
    """解析使用 timer/helper_cooldown 字段的单个分类"""
    results: list[UpgradeItem] = []
    now = datetime.now(timezone.utc)

    for item in raw_list:
        if not isinstance(item, dict):
            continue

        timer = _get_timer(item)
        if timer is None:
            continue

        data_id = item.get("data")
        lvl = _get_level(item)
        name = resolve_name(category, data_id)
        finish = now + timedelta(seconds=timer)

        results.append(UpgradeItem(
            category=category,
            item_name=name,
            item_level=lvl,
            timer_seconds=timer,
            finish_time=finish,
            data_id=data_id,
        ))

    return results


def parse_coc_json(raw_json: str | dict) -> list[UpgradeItem]:
    """
    解析 CoC 导出的 JSON 字符串或已解析的 dict

    支持所有升级类别:
      - buildings / spells / heroes / pets / equipment / units / helpers
      - buildings2 / heroes2 / units2 / siege_machines
    """
    if isinstance(raw_json, str):
        data = json.loads(raw_json)
    else:
        data = raw_json

    all_upgrades: list[UpgradeItem] = []

    timer_categories = [
        "buildings", "spells", "heroes", "pets", "equipment", "units",
        "helpers",
        "buildings2", "heroes2", "units2", "siege_machines",
    ]
    for cat in timer_categories:
        sub = data.get(cat)
        if sub and isinstance(sub, list):
            all_upgrades.extend(_parse_category_with_timer(sub, cat))

    return all_upgrades


def extract_player_info(raw_json: str | dict) -> dict:
    """提取玩家信息"""
    if isinstance(raw_json, str):
        try:
            data = json.loads(raw_json)
        except Exception:
            return {}
    else:
        data = raw_json

    # 提取玩家标签/名称
    player_tag = data.get("playerTag") or data.get("tag") or ""
    player_name = data.get("playerName") or data.get("name") or ""

    # 提取大本等级
    town_hall_level = None
    for bld in data.get("buildings", []):
        if isinstance(bld, dict) and bld.get("data") == 1000001:
            town_hall_level = bld.get("lvl", 0)
            break

    # 提取 Builder 数量 (从 builder_hut 1000015 的 cnt)
    builder_count = 0
    for bld in data.get("buildings", []):
        if isinstance(bld, dict) and bld.get("data") == 1000015:
            builder_count += bld.get("cnt", 0)

    return {
        "player_tag": player_tag,
        "player_name": player_name,
        "town_hall_level": town_hall_level,
        "builder_count": max(builder_count, 5) if builder_count == 0 else builder_count,
    }


def calculate_idle_times(upgrades: list[UpgradeItem]) -> dict:
    """计算工人/实验室空闲时间"""
    result = {
        "builder_idle_at": None,
        "lab_idle_at": None,
        "builder_busy_count": 0,
        "lab_busy_count": 0,
        "builder_total": None,
    }

    # 需要从原始 JSON 获取 builders 总数, 这里用占位
    building_items = [u for u in upgrades if u.category == "buildings"]
    spell_items = [u for u in upgrades if u.category == "spells"]

    if building_items:
        latest = max(building_items, key=lambda x: x.finish_time)
        result["builder_idle_at"] = latest.finish_time.isoformat()
        result["builder_busy_count"] = len(building_items)

    if spell_items:
        latest = max(spell_items, key=lambda x: x.finish_time)
        result["lab_idle_at"] = latest.finish_time.isoformat()
        result["lab_busy_count"] = len(spell_items)

    return result

File: shared/coc-parser.ts
/**
 * CoC JSON 本地解析器（移植自 backend/app/parser/coc_parser.py）
 * ============================================================
 * 解析逻辑：
 *  - JSON 中每个对象若有 "timer" 字段且 > 0，则正在升级
 *  - helpers 类别使用 "helper_cooldown" 字段
 *  - finish_time = 当前时间 + timer (秒)
 *  - 名称通过 coc-assets.ts 的 getItemNameById 解析为中文
 *
 * 纯前端，无后端依赖。Cloudflare Pages 静态部署可用。
 */

import { getItemNameById, inferCategory } from "./coc-assets";
import type { UpgradeItem, IdleTimes, PlayerInfo, VillageItem, VillageSnapshot } from "./types";
import { log, warn, error, group, fmtDuration } from "./logger";

const TIMER_CATEGORIES = [
  "buildings", "spells", "heroes", "pets", "equipment", "units",
  "helpers",
  "buildings2", "heroes2", "units2", "siege_machines",
] as const;

type RawItem = Record<string, unknown>;

function getTimer(item: RawItem): number | null {
  const timer = item.timer;
  if (typeof timer === "number" && timer > 0) return Math.floor(timer);

  const cooldown = item.helper_cooldown;
  if (typeof cooldown === "number" && cooldown > 0) return Math.floor(cooldown);

  return null;
}

function getLevel(item: RawItem): number {
  for (const field of ["lvl", "level"] as const) {
    const val = item[field];
    if (typeof val === "number") return val;
    if (typeof val === "string") {
      const n = parseInt(val, 10);
      if (!Number.isNaN(n)) return n;
    }
  }
  // 未找到等级字段，默认 1
  log("  getLevel: 未找到 lvl/level 字段，默认返回 1，item.data=", item.data);
  return 1;
}

function resolveName(category: string, dataId: number | null): string {
  if (dataId == null) return category;
  const info = getItemNameById(category, dataId);
  if (info) return info.zh;
  // 未知 SC ID — 资产表中找不到
  warn(`resolveName: 未知 SC ID ${dataId} (category=${category})，使用 ID 作为名称`);
  return String(dataId);
}

export interface ParseResult {
  upgrades: UpgradeItem[];
  player_info: PlayerInfo;
  idle_times: IdleTimes;
}

export function parseCocJson(raw: string | object, exportTime?: number): UpgradeItem[] {
  const inputType = typeof raw === "string" ? "string" : "object";
  const inputSize = typeof raw === "string" ? raw.length : JSON.stringify(raw).length;
  log(`parseCocJson: 开始解析，输入类型=${inputType}，大小=${inputSize} 字符`);

  let data: object;
  try {
    data = typeof raw === "string" ? JSON.parse(raw) : raw;
  } catch (e) {
    error("parseCocJson: JSON 解析失败", e);
    throw e;
  }

  const topKeys = Object.keys(data as Record<string, unknown>);
  log(`parseCocJson: 顶层字段 [${topKeys.join(", ")}]`);

  // 优先从 JSON 顶层 timestamp 字段读取导出时间（Unix 秒）
  // 其次用传入的 exportTime 参数，最后回退到当前时间
  const jsonObj = data as Record<string, unknown>;
  const jsonTs = typeof jsonObj.timestamp === "number" && jsonObj.timestamp > 0
    ? jsonObj.timestamp * 1000
    : null;
  const baseTime = jsonTs
    ?? (Number.isFinite(exportTime as number) ? (exportTime as number) : null)
    ?? Date.now();
  const diffMin = Math.round((Date.now() - baseTime) / 60000);
  log(`parseCocJson: 基准时间 ${new Date(baseTime).toLocaleString("zh-CN")} (距今 ${diffMin} 分钟)${jsonTs ? " [来自JSON timestamp]" : ""}`);

  const upgrades: UpgradeItem[] = [];
  let idCounter = 1;
  let totalSkipped = 0;

  for (const cat of TIMER_CATEGORIES) {
    const sub = (data as Record<string, unknown>)[cat];
    if (!Array.isArray(sub)) {
      log(`  [${cat}] 不存在或非数组，跳过`);
      continue;
    }

    let catActive = 0;
    let catSkipped = 0;

    for (const entry of sub) {
      if (typeof entry !== "object" || entry === null) {
        catSkipped++;
        continue;
      }
      const item = entry as RawItem;

      const timer = getTimer(item);
      if (timer == null) {
        catSkipped++;
        continue;
      }

      const dataId =
        typeof item.data === "number" ? item.data
          : typeof item.data === "string" ? parseInt(item.data, 10)
          : null;
      const lvl = getLevel(item);
      const name = resolveName(cat, dataId);
      const finishMs = baseTime + timer * 1000;

      log(`    → [${cat}] ${name} (scId=${dataId}, lvl=${lvl}, timer=${fmtDuration(timer)})`);

      upgrades.push({
        id: idCounter++,
        category: cat,
        item_name: name,
        item_level: lvl,
        timer_seconds: timer,
        finish_time: new Date(finishMs).toISOString(),
        notified: false,
        data_id: Number.isFinite(dataId as number) ? (dataId as number) : null,
      });
      catActive++;
    }

    log(`  [${cat}] 共 ${sub.length} 条，活跃升级 ${catActive}，跳过 ${catSkipped}`);
    totalSkipped += catSkipped;
  }

  log(`parseCocJson: 完成，共提取 ${upgrades.length} 个活跃升级，跳过 ${totalSkipped} 条无 timer 记录`);
  return upgrades;
}

export function extractPlayerInfo(raw: string | object): PlayerInfo {
  let data: Record<string, unknown>;
  try {
    data = (typeof raw === "string" ? JSON.parse(raw) : raw) as Record<string, unknown>;
  } catch (e) {
    error("extractPlayerInfo: JSON 解析失败，返回默认值", e);
    return {
      player_tag: "",
      player_name: "",
      town_hall_level: 0,
      builder_count: 5,
      active_upgrades: 0,
      completed_count: 0,
    };
  }

  const playerTag = (data.playerTag as string) || (data.tag as string) || "";
  const playerName = (data.playerName as string) || (data.name as string) || "";

  let townHallLevel = 0;
  let builderCount = 0;
  const buildings = data.buildings;
  if (Array.isArray(buildings)) {
    let foundTH = false;
    let foundBH = false;
    for (const b of buildings) {
      if (typeof b !== "object" || b === null) continue;
      const obj = b as RawItem;
      if (obj.data === 1000001) {
        townHallLevel = typeof obj.lvl === "number" ? obj.lvl : 0;
        foundTH = true;
        log(`  extractPlayerInfo: 检测到市政厅 Lv${townHallLevel}`);
      }
      if (obj.data === 1000015) {
        const cnt = typeof obj.cnt === "number" ? obj.cnt : 0;
        builderCount += cnt;
        foundBH = true;
        log(`  extractPlayerInfo: 检测到建筑工人小屋 cnt=${cnt}，累计=${builderCount}`);
      }
    }
    if (!foundTH) warn("extractPlayerInfo: 未找到市政厅 (data=1000001)，townHallLevel=0");
    if (!foundBH) warn("extractPlayerInfo: 未找到建筑工人小屋 (data=1000015)，使用默认值 5");
  } else {
    warn("extractPlayerInfo: buildings 字段不存在或非数组");
  }

  const result = {
    player_tag: playerTag,
    player_name: playerName,
    town_hall_level: townHallLevel,
    builder_count: builderCount > 0 ? builderCount : 5,
    active_upgrades: 0,
    completed_count: 0,
  };
  log(`extractPlayerInfo: 完成，玩家=${playerName || "(无名)"}，TH=${townHallLevel}，工人=${result.builder_count}`);
  return result;
}

export function calculateIdleTimes(upgrades: UpgradeItem[]): IdleTimes {
  const result: IdleTimes = {
    builder_idle_at: null,
    lab_idle_at: null,
    builder_busy_count: 0,
    lab_busy_count: 0,
    builder_total: null,
  };

  const buildingItems = upgrades.filter((u) => u.category === "buildings");
  const spellItems = upgrades.filter((u) => u.category === "spells");

  if (buildingItems.length > 0) {
    const latest = buildingItems.reduce((a, b) =>
      new Date(b.finish_time).getTime() > new Date(a.finish_time).getTime() ? b : a
    );
    result.builder_idle_at = latest.finish_time;
    result.builder_busy_count = buildingItems.length;
    log(`  calculateIdleTimes: 工人忙 ${buildingItems.length} 项，最晚完成=${latest.item_name} @ ${latest.finish_time}`);
  } else {
    log("  calculateIdleTimes: 无建筑升级，工人全部空闲");
  }

  if (spellItems.length > 0) {
    const latest = spellItems.reduce((a, b) =>
      new Date(b.finish_time).getTime() > new Date(a.finish_time).getTime() ? b : a
    );
    result.lab_idle_at = latest.finish_time;
    result.lab_busy_count = spellItems.length;
    log(`  calculateIdleTimes: 实验室忙 ${spellItems.length} 项，最晚完成=${latest.item_name} @ ${latest.finish_time}`);
  } else {
    log("  calculateIdleTimes: 无法术研究，实验室空闲");
  }

  return result;
}

export function parseFull(raw: string | object, exportTime?: number): ParseResult {
  group("parseFull 开始", () => {
    log(`输入类型=${typeof raw === "string" ? "string" : "object"}，大小=${typeof raw === "string" ? raw.length : JSON.stringify(raw).length} 字符`);
  });

  const upgrades = parseCocJson(raw, exportTime);
  const playerInfo = extractPlayerInfo(raw);
  playerInfo.active_upgrades = upgrades.filter(
    (u) => new Date(u.finish_time).getTime() > Date.now()
  ).length;
  playerInfo.completed_count = upgrades.length - playerInfo.active_upgrades;
  const idleTimes = calculateIdleTimes(upgrades);

  group("parseFull 汇总", () => {
    log(`活跃升级: ${playerInfo.active_upgrades}，已完成: ${playerInfo.completed_count}，总计: ${upgrades.length}`);
    log(`工人: 忙=${idleTimes.builder_busy_count}，空闲时间=${idleTimes.builder_idle_at || "(空闲)"}`);
    log(`实验室: 忙=${idleTimes.lab_busy_count}，空闲时间=${idleTimes.lab_idle_at || "(空闲)"}`);
  });

  return { upgrades, player_info: playerInfo, idle_times: idleTimes };
}

// ── 全量村庄解析（含 timer=0 的项）─────────────

/**
 * 解析整个村庄的所有条目（含 timer=0 的），生成 VillageSnapshot
 * - 与 parseFull() 互补：parseFull 只提取 timer>0 的活跃升级
 * - parseVillage 提取所有条目（含已完成/未升级的），用于基地分析/评分/推荐
 * - 不修改现有 parseCocJson/parseFull，向后兼容
 */
export function parseVillage(raw: string | object, exportTime?: number): VillageSnapshot {
  const inputSize = typeof raw === "string" ? raw.length : JSON.stringify(raw).length;
  log(`parseVillage: 开始全量解析，输入大小=${inputSize} 字符`);

  let data: Record<string, unknown>;
  try {
    data = typeof raw === "string" ? JSON.parse(raw) : (raw as Record<string, unknown>);
  } catch (e) {
    error("parseVillage: JSON 解析失败", e);
    throw e;
  }

  // 优先从 JSON 顶层 timestamp 字段读取导出时间（Unix 秒）
  const jsonTs = typeof data.timestamp === "number" && data.timestamp > 0
    ? data.timestamp * 1000
    : null;
  const baseTime = jsonTs
    ?? (Number.isFinite(exportTime as number) ? (exportTime as number) : null)
    ?? Date.now();
  log(`parseVillage: 基准时间 ${new Date(baseTime).toLocaleString("zh-CN")}${jsonTs ? " [来自JSON timestamp]" : ""}`);
  const items: VillageItem[] = [];

  let townHallLevel = 0;
  let builderCount = 0;
  const playerTag = (data.playerTag as string) || (data.tag as string) || "";
  const playerName = (data.playerName as string) || (data.name as string) || "";
  let unknownIdCount = 0;

  for (const cat of TIMER_CATEGORIES) {
    const sub = data[cat];
    if (!Array.isArray(sub)) continue;

    let catCount = 0;
    let catUpgrading = 0;

    for (const entry of sub) {
      if (typeof entry !== "object" || entry === null) continue;
      const item = entry as RawItem;

      const dataId =
        typeof item.data === "number" ? item.data
          : typeof item.data === "string" ? parseInt(item.data, 10)
          : null;
      if (!Number.isFinite(dataId as number)) {
        unknownIdCount++;
        continue;
      }
      const scId = dataId as number;

      // 市政厅等级
      if (scId === 1000001 && typeof item.lvl === "number") {
        townHallLevel = item.lvl;
        log(`  parseVillage: 检测到市政厅 Lv${townHallLevel}`);
      }
      // 建筑工人小屋数量
      if (scId === 1000015 && typeof item.cnt === "number") {
        builderCount += item.cnt;
        log(`  parseVillage: 检测到工人小屋 cnt=${item.cnt}，累计=${builderCount}`);
      }

      const currentLevel = getLevel(item);
      const timer = getTimer(item);

      const villageItem: VillageItem = {
        scId,
        category: cat,
        currentLevel,
        isUpgrading: timer != null,
      };

      if (timer != null) {
        villageItem.targetLevel = currentLevel + 1;
        villageItem.timerSeconds = timer;
        villageItem.finishTime = new Date(baseTime + timer * 1000).toISOString();
        catUpgrading++;
        log(`    → [${cat}] scId=${scId} lvl=${currentLevel}→${currentLevel + 1} timer=${fmtDuration(timer)}`);
      }

      items.push(villageItem);
      catCount++;
    }

    if (catCount > 0) {
      log(`  [${cat}] 共 ${catCount} 条，其中 ${catUpgrading} 条正在升级`);
    }
  }

  if (unknownIdCount > 0) {
    warn(`parseVillage: ${unknownIdCount} 条记录缺少有效 data 字段，已跳过`);
  }
  if (townHallLevel === 0) {
    warn("parseVillage: 未检测到市政厅等级 (data=1000001)，townHallLevel=0");
  }
  if (builderCount === 0) {
    warn("parseVillage: 未检测到建筑工人小屋 (data=1000015)，使用默认值 5");
  }

  const upgradingCount = items.filter((i) => i.isUpgrading).length;
  const snapshot: VillageSnapshot = {
    capturedAt: new Date(baseTime).toISOString(),
    townHallLevel,
    builderCount: builderCount > 0 ? builderCount : 5,
    playerTag,
    playerName,
    items,
  };

  group("parseVillage 完成", () => {
    log(`总条目: ${items.length}，正在升级: ${upgradingCount}，已满级/空闲: ${items.length - upgradingCount}`);
    log(`大本营: Lv${townHallLevel}，工人: ${snapshot.builderCount}`);
    log(`玩家: ${playerName || "(无名)"} (${playerTag || "(无tag)"})`);
  });

  return snapshot;
}

export { inferCategory };
